packages/grid/_modules_/grid/hooks/core/useGridStateInitialization.ts

```diff
@@ -1,33 +1,38 @@
 import React from 'react';
 import { DataGridProcessedProps } from '../../models/props/DataGridProps';
 import { GridApiRef } from '../../models/api/gridApiRef';
+import { GridApiCommon } from '../../models/api/gridApi';
 import { GridStateApi } from '../../models/api/gridStateApi';
 import { GridControlStateItem } from '../../models/controlStateItem';
 import { GridSignature } from '../utils/useGridApiEventHandler';
-import { GridState } from '../../models/gridState';
 import { GridEvents } from '../../models/events';
 import { useGridApiMethod } from '../utils';
-
-export const useGridStateInitialization = (apiRef: GridApiRef, props: DataGridProcessedProps) => {
-  const controlStateMapRef = React.useRef<Record<string, GridControlStateItem<any>>>({});
-  const [, rawForceUpdate] = React.useState<GridState>();
-
-  const updateControlState = React.useCallback<GridStateApi['unstable_updateControlState']>(
-    (controlStateItem) => {
-      const { stateId, ...others } = controlStateItem;
-
-      controlStateMapRef.current[stateId] = {
-        ...others,
-        stateId,
-      };
-    },
-    [],
-  );
-
-  const setState = React.useCallback<GridStateApi['setState']>(
+import { isFunction } from '../../utils/utils';
+
+export const useGridStateInitialization = <GridApi extends GridApiCommon>(
+  apiRef: GridApiRef<GridApi>,
+  props: DataGridProcessedProps,
+) => {
+  const controlStateMapRef = React.useRef<
+    Record<string, GridControlStateItem<GridApi['state'], any>>
+  >({});
+  const [, rawForceUpdate] = React.useState<GridApi['state']>();
+
+  const updateControlState = React.useCallback<
+    GridStateApi<GridApi['state']>['unstable_updateControlState']
+  >((controlStateItem) => {
+    const { stateId, ...others } = controlStateItem;
+
+    controlStateMapRef.current[stateId] = {
+      ...others,
+      stateId,
+    };
+  }, []);
+
+  const setState = React.useCallback<GridStateApi<GridApi['state']>['setState']>(
     (state) => {
-      let newState: GridState;
-      if (typeof state === 'function') {
+      let newState: GridApi['state'];
+      if (isFunction(state)) {
```

Comment:
`isFunction` tells TypeScript that the returned value is a function

```ts
export function isFunction(value: any): value is Function {
  return typeof value === 'function';
}
```

When using `typeof xxx === 'sometype'`, I don't know why but sometimes TypeScript narrows the typing after the check and sometime not.

And on this particular code, if I keep the `typeof state === 'function'`, TypeScript complains because it still thinkgs that `state` can be of type `GridState` in line 36.

Reply:
Well, good to know!